dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/compute/functions/CleanupServer.java

```diff
@@ -0,0 +1,153 @@
+import com.google.common.base.Predicate;
+import org.jclouds.compute.reference.ComputeServiceConstants;
+import org.jclouds.compute.reference.ComputeServiceConstants.Timeouts;
+import org.jclouds.dimensiondata.cloudcontrol.DimensionDataCloudControlApi;
+import org.jclouds.dimensiondata.cloudcontrol.domain.FirewallRule;
+import org.jclouds.dimensiondata.cloudcontrol.domain.NatRule;
+import org.jclouds.dimensiondata.cloudcontrol.domain.PublicIpBlock;
+import org.jclouds.dimensiondata.cloudcontrol.domain.Server;
+import org.jclouds.dimensiondata.cloudcontrol.domain.State;
+import org.jclouds.dimensiondata.cloudcontrol.features.NetworkApi;
+import org.jclouds.dimensiondata.cloudcontrol.features.ServerApi;
+import org.jclouds.dimensiondata.cloudcontrol.utils.DimensionDataCloudControlResponseUtils;
+import org.jclouds.logging.Logger;
+
+import javax.annotation.Resource;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import java.util.List;
+
+import static java.lang.String.format;
+import static org.jclouds.dimensiondata.cloudcontrol.utils.DimensionDataCloudControlResponseUtils.generateFirewallRuleName;
+import static org.jclouds.dimensiondata.cloudcontrol.utils.DimensionDataCloudControlResponseUtils.waitForServerState;
+
+@Singleton
+public class CleanupServer implements Function<String, Boolean> {
+
+   @Resource
+   @Named(ComputeServiceConstants.COMPUTE_LOGGER)
+   protected Logger logger = Logger.NULL;
+
+   private final DimensionDataCloudControlApi api;
+   private final Timeouts timeouts;
+
+   @Inject
+   public CleanupServer(final DimensionDataCloudControlApi api, final Timeouts timeouts) {
+      this.api = api;
+      this.timeouts = timeouts;
+   }
+
+   @Override
+   public Boolean apply(final String serverId) {
+      final ServerApi serverApi = api.getServerApi();
+      Server server = serverApi.getServer(serverId);
+
+      if (server == null) {
+         return true;
+      }
+
+      if (server.state().isFailed()) {
+         rollbackOperation(format("Server(%s) not deleted as it is in state(%s).", serverId, server.state()));
+      }
+
+      if (server.state().isNormal()) {
+
+         String networkDomainId = server.networkInfo().networkDomainId();
+         final String internalIp = server.networkInfo().primaryNic().privateIpv4();
+
+         // delete nat rules associated to the server, if any
+         final NetworkApi networkApi = api.getNetworkApi();
+         List<NatRule> natRulesToBeDeleted = networkApi.listNatRules(networkDomainId).concat()
+               .filter(new Predicate<NatRule>() {
+                  @Override
+                  public boolean apply(NatRule natRule) {
+                     return natRule.internalIp().equals(internalIp);
+                  }
+               }).toList();
+
+         for (final NatRule natRule : natRulesToBeDeleted) {
+            if (natRule.state().isNormal()) {
+               networkApi.deleteNatRule(natRule.id());
```

Comment:
Surround with a try/catch and log a warning, for a best-effort cleanup? Otherwise, it will stop deleting stuff here if this call fails. (Apply the same criteria to the other cleanup calls).

Reply:
Makes sense to cleanup as much as possible without failing the full operation, I will update to have this, but I am not sure about the need for a try/catch.

Reply:
Calls to the API can fail for a variety of reasons, say there is a conflict, a transient, error, whatever unexpected error that makes this raise an `HttpResponseException`. The `CleanupServer` is likely to be used when users want to destroy nodes and associated resources to "stop paying for them". We should be as defensive as possible and try to continue deleting stuff even if an unexpected exception occurred when trying to delete intermediate resources.

Reply:
Understood, thanks for clarifying. I will make that change.